### ISSUE TITLE
Supporting 24 bit stream types

### ISSUE THREAD

**Issue by MrMormon** [klingtnet/rsoundio#5]
Citing https://docs.klingt.net/rustdoc/rsoundio/enum.SioFormat.html:
```
Because there is no native number type that is 24 bits wide, (at least not in Rusts std) I had to use something like C unions in Rust to support them. This would probably be unsafe and working with 3 byte number types is also not too much fun.
```
[libsoundio's](http://libsound.io/doc-1.1.0/soundio_8h.html#aef491940cf81c91cf8ce15783a294af0) docs state that one passes 32 bit variables in this case, so the only reason I can see to need a special type at all is to simulate overflow. And if they're ever necessary, Rust has [plenty of bit operators](https://doc.rust-lang.org/book/syntax-index.html), which don't require unsafe.

**Comment by klingtnet** [comment]
So, the idea is to simulate a 24-bit value by masking a 32-bit integer?

**Comment by MrMormon** [comment]
Yep. The C lbrary isn't doing/requiring anything tricky, certainly nothing Rust can't handle.

**Comment by MrMormon** [comment]
What I don't know yet is whether the sample type's endianness affects the interpretation of Rust's primitive number types. I need to look at libsoundio's source, but I'm working on using it in Nim or Julia.

**Comment by klingtnet** [comment]
There is no special endianess handling, at least not in [`outstream_begin_write_alsa`](https://github.com/andrewrk/libsoundio/blob/ca0ebfda5c045dafc13bb1e22c2aa581fd94d6b7/src/alsa.c#L1444). Libsoundio set's the correct endian type for the format enums at compile time. We should take at look at every `outstream_begin_write` implementation in libsoundio to make sure that there is no special endianess handling.

**Comment by MrMormon** [comment]
I'm not confident I would know it was happening if I saw it.

**Comment by MrMormon** [comment]
I'm personally not worried about endianness for now, but andrewrk has an email address if we really want to be sure.

**Comment by klingtnet** [comment]
> Are you referring to this compile-time stuff?
Yes, that's what I was referring too.
Endianess shouldn't be a problem because almost all consumer machines nowadays are little endian ones.

**Comment by 0joshuaolson1** [comment]
I'm working with Andrew to add 24-bit packed/3bytes formats for the next major libsoundio version. The SoundIoFormat endianness seems to matter when writing to the buffer pointer at by a soundIoChannelArea[channel].ptr.

**Comment by klingtnet** [comment]
I am planning to make a rewrite of rsoundio for version `0.2` where I will address this issue.

**Comment by 0joshuaolson1** [comment]
libsoundio can use [24-bit](https://github.com/andrewrk/libsoundio/blob/master/soundio/soundio.h) formats now.

**Comment by klingtnet** [comment]
Thank your for the info, but there is no new libsoundio release, 1.1.0 is still the latest version.
I hope too find some time for the rsoundio rewrite because I am now working full time.
